From e666829949f82b3d8b7c722f7db5a8af38e93a2c Mon Sep 17 00:00:00 2001
From: Harald van Dijk <harald@gigawatt.nl>
Date: Tue, 10 Jan 2017 22:37:49 +0100
Subject: [PATCH] c++: Fix yyrestart(NULL) SEGV.

Binding a reference to a dereferenced null pointer is invalid and
compilers optimise away the &file == 0 check. We need a real stream.

yyin is available already, and yyrestart(NULL) is only supported when
yyin will not be used, so there is no harm in just passing in that.
Since we now always have a valid stream, we can skip the null check
too.

Fixes #98.
---
 src/flex.skl | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/flex.skl b/src/flex.skl
index 23e7063..66b7c1e 100644
--- a/src/flex.skl
+++ b/src/flex.skl
@@ -1896,6 +1896,9 @@ m4_ifdef( [[M4_YY_USE_LINENO]],
  */
 void yyFlexLexer::yyrestart( std::istream* input_file )
 {
+	if( ! input_file ) {
+		input_file = &yyin;
+	}
 	yyrestart( *input_file );
 }
 %endif
@@ -2057,7 +2060,7 @@ static void yy_load_buffer_state  (M4_YY_DEF_ONLY_ARG)
 	b->yy_input_file = file;
 %endif
 %if-c++-only
-	b->yy_input_file = (&file == 0) ? NULL : file.rdbuf();
+	b->yy_input_file = file.rdbuf();
 %endif
 	b->yy_fill_buffer = 1;
 
